File: openwrt.go
// +build openwrt

package main

import (
	"bytes"
	"fmt"
	"net"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"unicode"

	"github.com/aymerick/raymond"
)

const backend = "openwrt"

type (
	netInterface struct {
		ifname   []string
		dhcp     bool
		typ      string
		ipAddres *ip
	}
	emitter struct {
		devices    map[string]struct{}
		interfaces map[string]netInterface
		wireless   *bytes.Buffer
		network    *bytes.Buffer
		tplout     map[string]string
	}
)

func newEmitter() emitter {
	return emitter{
		devices:    make(map[string]struct{}, 0),
		interfaces: make(map[string]netInterface, 0),
		wireless:   &bytes.Buffer{},
		network:    &bytes.Buffer{},
		tplout:     make(map[string]string, 0),
	}
}

func (e *emitter) load(config genesis) error {
	for name, dev := range config.Devices {
		e.loadDevice(name, dev)
	}
	for name, devIntf := range config.Interfaces {
		e.loadDeviceInterface(name, devIntf)
	}
	for name, tmpl := range config.Templates {
		e.loadTemplate(name, tmpl)
	}

	_, err := fmt.Fprintf(e.network, `
config interface   'loopback'
	option ifname  'lo'
	option proto   'static'
	option ipaddr  '127.0.0.1'
	option netmask '255.0.0.0'

`)
	if err != nil {
		return err
	}

	for k, v := range e.interfaces {
		_, err = fmt.Fprintf(e.network, "config interface   '%s'\n", k)
		if err != nil {
			return err
		}
		if v.typ != "" {
			_, err = fmt.Fprintf(e.network, "	option type    '%s'\n", v.typ)
			if err != nil {
				return err
			}
		}

		if len(v.ifname) != 0 {
			_, err = fmt.Fprintf(e.network, "	option ifname  '%s'\n", strings.Join(v.ifname, " "))
			if err != nil {
				return err
			}
		}

		if v.dhcp {
			_, err = fmt.Fprintf(e.network, "	option proto   'dhcp'\n")
			if err != nil {
				return err
			}
		} else if v.ipAddres != nil {
			netmask := net.IP(v.ipAddres.Mask)
			_, err = fmt.Fprintf(e.network, `	option proto   'static'
	option ipaddr  '%s'
	option netmask '%s'
`, v.ipAddres.IP, netmask)
			if err != nil {
				return err
			}
		}
		_, err = fmt.Fprintln(e.network, "")
		if err != nil {
			return err
		}
	}
	return nil
}

func (e *emitter) loadDeviceInterface(name string, devIntf devInterface) error {
	for _, r := range name {
		if !unicode.IsLetter(r) && (48 > r || r > 58) {
			return fmt.Errorf("invalid interface name %s", name)
		}
	}
	if name == "" {
		panic(devIntf)
	}
	switch devIntf.Class {
	case "":
		e.createDeviceInterface(name, devIntf)
	case "bridge":
		e.createDeviceInterface(name, devIntf)
		intf := e.interfaces[name]
		intf.typ = "bridge"
		e.interfaces[name] = intf
	case "wifi":
		return e.createWifi(name, devIntf)
	case "wireguard":
		return e.createWireguard(name, devIntf)
	default:
		return fmt.Errorf("invalid class '%s'", devIntf.Class)

	}
	return nil
}

func (e *emitter) createDeviceInterface(name string, devIntf devInterface) error {
	if devIntf.Bridge != "" && devIntf.Device != "" {
		intf := e.interfaces[devIntf.Bridge]
		intf.ifname = append(intf.ifname, devIntf.Device)
		e.interfaces[devIntf.Bridge] = intf
		return nil
	}
	intf := e.interfaces[name]

	if devIntf.Device != "" {
		intf.ifname = append(intf.ifname, devIntf.Device)
	}

	if len(devIntf.IPAddres) != 0 {
		addr, err := ipFromBytes([]byte(devIntf.IPAddres[0]))
		if err != nil {
			return err
		}
		intf.ipAddres = addr
	}

	switch dhcp := devIntf.Dhcp.(type) {
	case string:
		switch dhcp {
		case "yes", "true", "ipv4":
			intf.dhcp = false
		}
	case bool:
		intf.dhcp = dhcp
	}
	e.interfaces[name] = intf
	return nil
}

func (e *emitter) createWifi(name string, devIntf devInterface) error {
	if devIntf.Device == "" {
		return fmt.Errorf("wifi interface missing device '%s'", name)
	}
	if _, ok := e.devices[devIntf.Device]; ok {
		return fmt.Errorf("undeclared interface '%s' used in device '%s'", devIntf.Device, name)
	}
	var mode string
	switch devIntf.Wifi.Mode {
	case sta:
		mode = "sta"
	case ap:
		mode = "ap"
	case monitor:
		mode = "monitor"
	}
	_, err := fmt.Fprintf(e.wireless, `
config wifi-iface   '%s'
	option device   '%s'
	option mode     '%s'
	option ifname   '%s'
`, name, devIntf.Device, mode, name)
	if err != nil {
		return err
	}

	if devIntf.Bridge != "" {
		_, err = fmt.Fprintf(e.wireless, "	option network  '%s'\n", devIntf.Bridge)
		if err != nil {
			return err
		}
	}
	if devIntf.Wifi.Ssid != "" {
		_, err = fmt.Fprintf(e.wireless, "	option ssid     '%s'\n", devIntf.Wifi.Ssid)
		if err != nil {
			return err
		}

		switch devIntf.Wifi.Auth {
		case psk2:
			_, err = fmt.Fprint(e.wireless, "	option encryption 'psk2' \n")
		case none:
			_, err = fmt.Fprint(e.wireless, "	option encryption 'none' \n")
		}
		if err != nil {
			return err
		}

	}
	if devIntf.Wifi.Key != "" {
		_, err = fmt.Fprintf(e.wireless, "	option key       '%s'\n", devIntf.Wifi.Key)
		if err != nil {
			return err
		}

	}
	return nil
}

func (e *emitter) createWireguard(name string, devIntf devInterface) error {
	if devIntf.Wireguard.PrivateKey == "" {
		return fmt.Errorf("missing [interface.%s.wireguard] section", name)
	}
	_, err := fmt.Fprintf(e.network, `
config interface '%s'
	option proto 'wireguard'
	option private_key '%s'
	option disabled '0'
`, name, devIntf.Wireguard.PrivateKey)
	if err != nil {
		return err
	}

	for _, addr := range devIntf.IPAddres {
		_, err = fmt.Fprintf(e.network, "	list addresses '%v'\n", addr)
		if err != nil {
			return err
		}
	}

	if len(devIntf.Wireguard.Peers) != 1 {
		return fmt.Errorf("need exactly one [interface.%s.wireguard.peers] section", name)
	}
	peer := devIntf.Wireguard.Peers[0]
	autoroute := "0"
	if peer.Autoroute {
		autoroute = "1"
	}

	_, err = fmt.Fprintf(e.network, `
config wireguard_%s
	option public_key '%s'
	option route_allowed_ips '%s'
`, name, peer.PublicKey, autoroute)
	if err != nil {
		return err
	}

	if peer.Psk != "" {
		_, err = fmt.Fprintf(e.network, "	option preshared_key '%v'\n", peer.Psk)
		if err != nil {
			return err
		}
	}

	endpoints := strings.Split(peer.Endpoint, ":")
	if len(endpoints) != 2 {
		return fmt.Errorf("invalid wg endpoint: %s", peer.Endpoint)
	}
	endpointPort, err := strconv.Atoi(endpoints[1])
	if err != nil {
		return fmt.Errorf("invalid wg endpoint port: %s: %s", endpoints[1], err)
	}

	_, err = fmt.Fprintf(e.network, `	option endpoint_port '%d'
	option endpoint_host
'%s'`, endpointPort, endpoints[0])
	if err != nil {
		return err
	}

	if peer.Keepalive != nil {
		_, err = fmt.Fprintf(e.network, "	option persistent_keepalive '%d'\n", *peer.Keepalive)
		if err != nil {
			return err
		}
	}

	for _, route := range peer.Routes {
		_, err = fmt.Fprintf(e.network, "	list allowed_ips '%s'\n", route)
		if err != nil {
			return err
		}
	}

	return nil
}

func (e *emitter) loadDevice(name string, dev device) error {
	e.devices[name] = struct{}{}
	switch dev.Class {
	case "wifi":
		if dev.Path == "" {
			return fmt.Errorf("wifi device must be matched by path %s", name)
		}
		path := dev.Path
		if !strings.HasPrefix(dev.Path, "/sys/devices/") {
			path, _ = filepath.EvalSymlinks(filepath.Join(dev.Path, "device"))
			//if the error != nil, path will be empty and thus will fail the next check.
		}
		if !strings.HasPrefix(path, "/sys/devices/") {
			return fmt.Errorf("path %s did not resolve to a sysfs device", path)
		}

		path = strings.TrimPrefix(path, "/sys/devices/")

		_, err := fmt.Fprintf(e.wireless, `
config wifi-device '%s'
	option type     'mac80211'
	option path     '%s'
`, name, path)
		if err != nil {
			return err
		}

		if dev.Wifi == nil {
			return fmt.Errorf("missing [device.%s.wifi] section", name)
		}
		//all we need to do is make sure its one of the suported input types.
		//fmt will format strings and ints for us.
		_, ok1 := dev.Wifi.Channel.(int)
		_, ok2 := dev.Wifi.Channel.(string)
		if ok1 || ok2 {
			_, err = fmt.Fprintf(e.wireless, "	option channel  '%v'\n", dev.Wifi.Channel)
			if err != nil {
				return err
			}
		}

		if dev.Wifi.Htmode != "" {
			_, err = fmt.Fprintf(e.wireless, "	option htmode  '%v'\n", dev.Wifi.Htmode)
			if err != nil {
				return err
			}
		}
	default:
		return fmt.Errorf("invalid class '%s'", dev.Class)

	}
	return nil
}

func (e *emitter) loadTemplate(name string, tmpl template) error {
	handlebar, err := raymond.ParseFile(filepath.Join("./etc/devguard/genesis/templates/", tmpl.Template))
	if err != nil {
		return err
	}
	res, err := handlebar.Exec(tmpl.Variables)
	if err != nil {
		return err
	}
	e.tplout[tmpl.Output] = res
	return nil
}

func (e *emitter) commit() error {
	err := os.MkdirAll("./etc/config", 0700)
	if err != nil {
		return err
	}
	file, err := os.Create("./etc/config/wireless")
	if err != nil {
		return err
	}
	e.wireless.WriteTo(file)
	file.Close()
	file, err = os.Create("./etc/config/network")
	if err != nil {
		return err
	}
	e.network.WriteTo(file)
	file.Close()
	for path, data := range e.tplout {
		file, err = os.Create(path)
		if err != nil {
			return err
		}
		file.WriteString(data)
	}

	return exec.Command("./etc/devguard/genesis/post").Run()
}

File: src/openwrt.rs
use super::genesis;
use std::io::{Write, Read};
use std::io::{Error,ErrorKind};
use std::collections::HashMap;
use ipnet::IpNet;
use std::process::Command;
use handlebars::Handlebars;

#[derive(Default)]
struct Interface {
    ifname:     Vec<String>,
    dhcp:       bool,
    typ:        Option<String>,
    ipaddr:     Option<(String, String)>,
    gateway:    Option<String>,
    dns:        Option<Vec<String>>,
}

pub struct Device {
}

pub struct Emitter{
    devices:        HashMap<String, Device>,
    interfaces:     HashMap<String, Interface>,
    out_wireless:   Vec<u8>,
    out_network:    Vec<u8>,
    tplout:         HashMap<String, String>,
}

impl Emitter {
    pub fn new() -> Self {
        Self {
            devices:        HashMap::new(),
            interfaces:     Default::default(),
            out_wireless:   Vec::new(),
            out_network:    Vec::new(),
            tplout:         HashMap::new(),
        }
    }


    pub fn load(&mut self, config: &genesis::Genesis) -> Result<(), Error> {
        for (name, device ) in &config.device {
            self.device(name, &device)?;
        }

        for (name, interface) in &config.interface {
            self.interface(name, &interface)?;
        }


        for (name, template) in &config.template {
            self.template(name, &template)?;
        }

        write!(&mut self.out_network, "
config interface   'loopback'
    option ifname  'lo'
    option proto   'static'
    option ipaddr  '127.0.0.1'
    option netmask '255.0.0.0'
\n")?;

        //emit
        for (k,v) in &self.interfaces {
            write!(&mut self.out_network, "config interface   '{}'\n", k)?;
            if let Some(typ) = &v.typ{
                write!(&mut self.out_network,"    option type    '{}'\n", typ)?;
            }
            if v.ifname.len() > 0 {
                write!(&mut self.out_network,"    option ifname  '{}'\n", v.ifname.join(" "))?;
            }

            if v.dhcp {
                write!(&mut self.out_network,"    option proto   'dhcp'\n")?;
            }
            if let Some((ipaddr, mask)) = &v.ipaddr {
                write!(&mut self.out_network,"    option proto   'static'\n")?;
                write!(&mut self.out_network,"    option ipaddr  '{}'\n",  ipaddr)?;
                write!(&mut self.out_network,"    option netmask '{}'\n", mask)?;
            }

            if let Some(gw) = &v.gateway{
                write!(&mut self.out_network,"    option gateway '{}'\n", gw)?;
            }
            if let Some(dns) = &v.dns{
                for dns in dns {
                    write!(&mut self.out_network,"    list dns '{}'\n", dns)?;
                }
            }

            write!(&mut self.out_network, "\n")?;
        }


        Ok(())
    }

    fn device(&mut self, name: &str, device: &genesis::Device) -> Result<(),Error> {
        self.devices.insert(name.to_string(), Device{
        });
        match device.class.as_str() {
            "wifi" => {

                let mut path = match &device.path {
                    Some(v) => v.clone(),
                    None => return Err(Error::new(ErrorKind::Other, format!("wifi device must be matched by path '{}'", name))),
                };

                if !path.starts_with("/sys/device") {
                    let path2 = std::path::Path::new(&path);
                    if let Ok(v) = std::fs::read_link(&path2.join("device")).and_then(|v|std::path::Path::new(&path).join(v).canonicalize()) {
                        path = v.clone().to_string_lossy().to_string();
                    }
                }

                if !path.starts_with("/sys/devices/") {
                    return Err(Error::new(ErrorKind::Other, format!("path {} did not resolve to a sysfs device", path)));
                }
                path = path[13..].to_string();


                write!(&mut self.out_wireless , "
config wifi-device '{}'
    option type     'mac80211'
    option path     '{}'
", name, &path)?;

                let w = match device.wifi.as_ref() {
                    Some(v) => v,
                    None => {
                        return Err(Error::new(ErrorKind::Other, format!("missing [device.{}.wifi] section", name)));
                    }
                };

                match &w.channel {
                    Some(toml::value::Value::String(v)) =>  {
                        write!(&mut self.out_wireless , "    option channel  '{}'\n", v)?;
                    }
                    Some(toml::value::Value::Integer(v)) =>  {
                        write!(&mut self.out_wireless , "    option channel  '{}'\n", v)?;
                    },
                    _ => ()
                };

                if let Some(v) = &w.htmode {
                    write!(&mut self.out_wireless , "    option htmode  '{}'\n", v)?;
                }

            },
            a => {
                return Err(Error::new(ErrorKind::Other, format!("invalid class '{}'", a)));
            }
        }
        Ok(())
    }

    fn interface(&mut self, name: &str, interface: &genesis::Interface) -> Result<(),Error> {
        if name.chars().any(|x| !x.is_alphanumeric()) {
            return Err(Error::new(ErrorKind::Other, format!("invalid interface name {}", name)));
        }

        match interface.class.as_ref().map(|x|x.as_ref()) {
            None => {
                return self.create_interface(name, interface);
            },
            Some("bridge") => {
                self.create_interface(name.clone(),interface)?;
                self.interfaces.get_mut(name).unwrap().typ = Some("bridge".to_string());
            },
            Some("wifi") => {
                if interface.bridge.is_none() {
                    let mut interface = interface.clone();
                    interface.device = None;
                    self.create_interface(name.clone(), &interface)?;
                }
                return self.wifi(name, interface);
            },
            Some("wireguard") => {
                self.wireguard(name, interface)?;
            },
            Some(a) => {
                return Err(Error::new(ErrorKind::Other, format!("invalid class '{}'", a)));
            }
        }
        Ok(())
    }

    fn wifi(&mut self, name: &str, interface: &genesis::Interface) -> Result<(),Error> {

        let device_name = match &interface.device {
            None => return Err(Error::new(ErrorKind::Other, format!("wifi interface missing device {}'", name))),
            Some(v) => match self.devices.get(v) {
                None => return Err(Error::new(ErrorKind::Other, format!("undeclared interface '{}' used in device {}'", v, name))),
                Some(_) => v.to_string(),
            }
        };

        let w = match interface.wifi.as_ref() {
            Some(v) => v,
            None => {
                return Err(Error::new(ErrorKind::Other, format!("missing [interface.{}.wifi] section", name)));
            }
        };

        let mode = match w.mode {
            genesis::WifiMode::Sta      => "sta",
            genesis::WifiMode::Ap       => "ap",
            genesis::WifiMode::Monitor  => "monitor",
        };

        write!(&mut self.out_wireless , "
config wifi-iface   '{}'
    option device   '{}'
    option mode     '{}'
    option ifname   '{}'
", name, device_name, mode, name)?;

        if let Some(br) = &interface.bridge {
            write!(&mut self.out_wireless , "    option network  '{}'\n", br)?;
        } else {
            write!(&mut self.out_wireless , "    option network  '{}'\n", name)?;
        }

        if let Some(v) = &w.ssid {
            write!(&mut self.out_wireless , "    option ssid     '{}'\n", v)?;

            match &w.auth {
                Some(genesis::WifiAuth::Psk2) => {
                    write!(&mut self.out_wireless , "    option encryption 'psk2' \n")?;
                },
                Some(genesis::WifiAuth::None) | None  => {
                    write!(&mut self.out_wireless , "    option encryption 'none' \n")?;
                }
            };
        };

        if let Some(v) = &w.key {
            write!(&mut self.out_wireless , "    option key       '{}'\n", v)?;
        };

        Ok(())
    }

    fn create_interface(&mut self, name: &str, interface: &genesis::Interface) -> Result<(),Error> {

        if let Some(br) = &interface.bridge {
            let i = self.interfaces.entry(br.to_string()).or_default();
            if let Some(ifname) = &interface.device {
                i.ifname.push(ifname.to_string());
            }
            return Ok(());
        }

        let i = self.interfaces.entry(name.to_string()).or_default();
        if let Some(ifname) = &interface.device {
            i.ifname.push(ifname.to_string());
        }

        if let Some(ipaddr) = &interface.ipaddr {
            if let Some(addr) = ipaddr.first() {
                let ip : IpNet = addr.parse().map_err(super::map_err)?;
                i.ipaddr = Some((format!("{}", ip.addr()), format!("{}", ip.netmask())));
            }
        }

        i.gateway   = interface.gateway.clone();
        i.dns       = interface.dns.clone();

        match &interface.dhcp {
            Some(toml::value::Value::String(s)) =>  {
                match s.as_str() {
                    "yes" | "true" | "ipv4" => {
                        i.dhcp = true;
                    },
                    "false" | "no" => {
                        i.dhcp = false;
                    },
                    _ => (),
                }
            },
            Some(toml::value::Value::Boolean(v)) => {
                if *v {
                    i.dhcp = true;
                } else {
                    i.dhcp = true;
                }
            },
            _ => (),
        }
        Ok(())
    }


    fn wireguard(&mut self, name: &str, interface: &genesis::Interface) -> Result<(),Error> {
        let wg = match interface.wireguard.as_ref() {
            Some(v) => v,
            None => {
                return Err(Error::new(ErrorKind::Other, format!("missing [interface.{}.wireguard] section", name)));
            }
        };
        write!(&mut self.out_network, "
config interface '{}'
    option proto 'wireguard'
    option private_key '{}'
    option disabled '0'
", name, wg.private_key)?;

        if let Some(ipaddr) = &interface.ipaddr {
            for addr in ipaddr {
                write!(&mut self.out_network, "    list addresses '{}'\n", addr)?;
            }
        }

        let peer = match &wg.peers {
            Some(v) if v.len() == 1 => v.first().unwrap(),
            _ => {
                return Err(Error::new(ErrorKind::Other, format!("need exactly one [interface.{}.wireguard.peers] section", name)));
            }
        };


        write!(&mut self.out_network, "
config wireguard_{}
    option public_key '{}'
    option route_allowed_ips '{}'
", name, peer.public_key, if peer.autoroute.clone().unwrap_or_default() {"1"} else {"0"})?;

        if let Some(psk) = &peer.psk {
            write!(&mut self.out_network, "    option preshared_key '{}'\n", psk)?;
        }

        let endpoint : Vec<&str> = peer.endpoint.split(":").collect();
        if endpoint.len() != 2 {
            return Err(Error::new(ErrorKind::Other, format!("invalid wg endpoint: {}", peer.endpoint)));
        }
        match endpoint[1].parse::<u16>() {
            Err(e) => {
                return Err(Error::new(ErrorKind::Other, format!("invalid wg endpoint port: {}: {}", endpoint[1], e)));
            }
            Ok(v) => {
                write!(&mut self.out_network, "    option endpoint_port '{}'\n", v)?;
            },
        };
        write!(&mut self.out_network, "    option endpoint_host '{}'\n", endpoint[0])?;


        if let Some(v) = &peer.keepalive {
            write!(&mut self.out_network, "    option persistent_keepalive '{}'\n", v)?;
        }

        for route in &peer.routes {
            write!(&mut self.out_network, "    list allowed_ips '{}'\n", route)?;
        }

        Ok(())
    }

    fn template(&mut self, _name: &str, template: &genesis::Template) -> Result<(),Error> {
        let p = std::path::Path::new("/etc/devguard/genesis/templates/").join(&template.template);
        let mut f = std::fs::File::open(p)?;
        let mut s = String::new();
        f.read_to_string(&mut s)?;

        let t = Handlebars::new().render_template(&s, &template.vars).map_err(super::map_err)?;

        self.tplout.insert(template.output.to_string(), t);

        Ok(())
    }

    pub fn commit(&self) ->  Result<(), Error> {

        let mut f = std::fs::File::create("/etc/config/wireless")?;
        f.write_all(&self.out_wireless)?;

        let mut f = std::fs::File::create("/etc/config/network")?;
        f.write_all(&self.out_network)?;


        for (path, s) in &self.tplout {
            if let Err(err) = std::fs::File::create(path).and_then(|mut f|f.write_all(s.as_bytes())) {
                log::warn!("{}: {}", path, err);
            }
        }

        Command::new("/etc/devguard/genesis/post")
            .spawn().ok();

        Ok(())
    }



}


